src/webots/nodes/WbBallJoint.cpp

```diff
@@ -74,64 +323,173 @@ bool WbBallJoint::setJoint() {
+  if (parentBody == NULL && body == NULL)
+    dJointDisable(mControlMotor);
+  else
+    dJointEnable(mControlMotor);
+  setOdeJoint(body, parentBody);
 
   return true;
 }
 
-void WbBallJoint::setOdeJoint(dBodyID body, dBodyID parentBody) {
-  WbBasicJoint::setOdeJoint(body, parentBody);
-  applyToOdeAnchor();
+double WbBallJoint::position(int index) const {
+  switch (index) {
+    case 1:
+      return mPosition;
+    case 2:
+      return mPosition2;
+    case 3:
+      return mPosition3;
+    default:
+      return NAN;
+  }
+}
+
+double WbBallJoint::initialPosition(int index) const {
+  switch (index) {
+    case 1:
+      return mInitialPosition;
+    case 2:
+      return mInitialPosition2;
+    case 3:
+      return mInitialPosition3;
+    default:
+      return NAN;
+  }
+}
+
+void WbBallJoint::setPosition(double position, int index) {
+  if (index == 3) {
+    mPosition3 = position;
+    mOdePositionOffset3 = position;
+    WbJointParameters *const p3 = parameters3();
+    if (p3)
+      p3->setPosition(mPosition3);
+
+    WbMotor *const m3 = motor3();
+    if (m3)
+      m3->setTargetPosition(position);
+    return;
+  }
+  WbHinge2Joint::setPosition(position, index);
+}
+
+bool WbBallJoint::resetJointPositions() {
+  mOdePositionOffset3 = 0.0;
+  return WbHinge2Joint::resetJointPositions();
+}
+
+void WbBallJoint::updateOdePositionOffset() {
+  WbHinge2Joint::updateOdePositionOffset();
+  mOdePositionOffset3 = position(3);
 }
 
 void WbBallJoint::preFinalize() {
-  WbBasicJoint::preFinalize();
+  WbHinge2Joint::preFinalize();
 
   WbBallJointParameters *const p = ballJointParameters();
   if (p && !p->isPreFinalizedCalled())
     p->preFinalize();
+
+  for (int i = 0; i < devices3Number(); ++i) {
+    if (device3(i) && !device3(i)->isPreFinalizedCalled())
+      device3(i)->preFinalize();
+  }
+  // the following code could be simplified
+  WbBaseNode *const p3 = dynamic_cast<WbBaseNode *>(mParameters3->value());
+  if (p3 && !p3->isPreFinalizedCalled())
+    p3->preFinalize();
+
+  updateParameters3();
+  checkMotorLimit();
+
+  mInitialPosition3 = mPosition3;
 }
 
 void WbBallJoint::postFinalize() {
-  WbBasicJoint::postFinalize();
+  WbHinge2Joint::postFinalize();
 
   WbBallJointParameters *const p = ballJointParameters();
 
   if (p && !p->isPostFinalizedCalled())
     p->postFinalize();
-}
-
-void WbBallJoint::updateAnchor() {
-  if (mJoint)
-    applyToOdeAnchor();
-
-  if (WbWrenRenderingContext::instance()->isOptionalRenderingEnabled(WbWrenRenderingContext::VF_JOINT_AXES))
-    updateJointAxisRepresentation();
-}
-
-void WbBallJoint::applyToOdeAnchor() {
-  const WbMatrix4 &m4 = upperTransform()->matrix();
-  const WbVector3 &t = m4 * anchor();
-  dJointSetBallAnchor(mJoint, t.x(), t.y(), t.z());
+  for (int i = 0; i < devices3Number(); ++i) {
+    if (device3(i) && !device3(i)->isPostFinalizedCalled())
+      device3(i)->postFinalize();
+  }
+  // the following code could be simplified
```

Comment:
Why don't we simplify it now?

Reply:
Because I have no idea how we could simplify this, but since the comment is present for the Hinge2Joint it probably applies here too.

Reply:
Ok, then the comment should be either more elaborated or removed (if we give up on simplification).